File: my_code/DSA/JS/DS/stack and queue/queue.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

// best to add to the last and remove from first
// we don't need pop its O(n)
// so will use push, and shift => but call it pop
class Queue {
  constructor() {
    this.first = null;
    this.last = null;
    this.size = 0;
  }

  // push add to end O(1)
  // called enqueue
  enqueue(val) {
    // make a new node to push
    const newNode = new Node(val);
    if (!this.first) {
      this.first = newNode;
      this.last = newNode;
    } else {
      // set the last node next to point into the new node
      this.last.next = newNode;
      //  and set the last node to be the new node itself
      this.last = newNode;
    }
    return ++this.size;
  }

  // shift remove from first O(1)
  // called dequeue
  dequeue() {
    if (!this.first) {
      return null;
    }
    // store the last in (first node) to return after removing
    const poppedNode = this.first;

    // if the size is one just one node
    if (this.first === this.last) {
      // reset the initial values
      this.first = null;
      this.last = null;
    } else {
      // set the first to be the new first
      this.first = this.first.next;
    }
    this.size--;
    return poppedNode.val;
  }
}

File: my_code/DSA/PYTHON/DS/stack and queue/queue.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

class Queue:
    def __init__(self):
        self.first = None
        self.last = None
        self.size = 0

    # push add to end O(1)
    # called enqueue
    def enqueue(self, val):
        # make a new node to push
        new_node = Node(val)

        if not self.first:
            self.first = new_node
            self.last = new_node
        else:
            # set the last node next to point into the new node
            self.last.next = new_node
            # and set the last node to be the new node itself
            self.last = new_node
        self.size += 1
        return self.size
    
    # shift remove from first O(1)
    # called dequeue
    def dequeue(self):
        if not self.first:
            return None
        
        # store the last in (first node) to return after removing
        popped_node = self.first

        # if the size is one just one node
        if self.size == 1:
            # reset the initial values
            self.first = None
            self.last = None
        else:
            # set the first to be the new first
            self.first = self.first.next
        self.size -= 1
        return popped_node.val